ykcapi/src/lib.rs

```diff
@@ -28,6 +28,12 @@ pub extern "C" fn yk_control_point(mt: *mut MT, loc: *mut Location) {
     }
 }
 
+#[no_mangle]
+pub extern "C" fn yk_stopgap_safe() {
+    // This function is a marker for the trace compiler. It might need to become an LLVM builtin,
+    // or we may need to add an LLVM annotation or ...
+}
```

Comment:
LLVM has ways to go from compiled code to an interpreter, but I can't find anything to go from an interpreter to compiled code.

https://llvm.org/docs/StackMaps.html, https://llvm.org/docs/LangRef.html#llvm-experimental-deoptimize-intrinsic

Reply:
The "new" yk (which, unlike "old" yk isn't fundamentally Rust based) is, in a sense, the missing bit that you're referring to. It (will) use things like stackmaps and `deoptimize` to help it do so. That does raise some interesting challenges, and one thing that's implicit in this PR is that I think we'll need to extend clang/LLVM to support function attributes that yk will need. This is all a bit sketch-y at the moment, so the precise details will probably alter!

Reply:
I mean that LLVM currently doesn't have a way to setup the stackframe and jump into the middle of a function AFAICT. It only has support for extracting info out of the stackframe and jumping out of the middle of a function into an interpreter. Stackmaps store where variables can be read, but they may very well also be held in registers at the same time with the same value or derived values may be held in registers (or on the stack without the stackmap mentioning how to compute them or mentioning them at all). Jumping into the middle of a function would require computing the derived values and storing the right value in registers.

```rust
fn foo(a: u8) -> u8 {
    safepoint();
    a * 3
}
```

could be transformed into for example

```rust
fn foo(a: u8) -> u8 {
    let tmp = a << 1;
    safepoint();
    a + tmp
}
```

The stackmap would only record `a` at the safepoint, however to safely jump into `foo` at the safepoint, you have to store `tmp` at the right location. Another example would be

```rust
fn foo(a: u8) -> u8 {
    foo(&a);
    safepoint();
    a
}
```

could be codegened as

```rust
fn foo(a: u8) -> u8 { // a is stored in a register
    // store a on the stack here
    foo(&a); // reference to the a on the stack is passed to foo
    safepoint(); // stackmap references a on the stack
    a // return the version of a in a register
}
```

here only the a on the stack is marked in the stackmap, but the a in a register is actually used for the return value and as such needs to be written when jumping into foo at the safepoint.

Stackmaps are sufficient for reading the value of all locals, but not for writing it as required to jump into the middle of a function from the interpreter.

Reply:
That's my understanding to. Basically we expect control flow to look like:

```c
pc = 0;
while (true) {
   control_point(...);
   if opcodes[pc] == PUSH {...}
   else if opcodes[pc] == BACKWARDS_JUMP {
      pc -= val(opvalue[pc]);
      yk_stopgap_safe();
      continue;
  }
}
```

As a good approximation of the truth, you go into compiled code at `control_point` and come out of it at `stopgap_safe`: the latter thus needs to put the frame in the "right state" for `control_point` to take back over. This is far beyond what any C spec is likely to declare as defined behaviour, but there's no alternative that I can think of.